subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java

```diff
@@ -233,54 +235,44 @@ private void defineConfigurationsForSourceSet(SourceSet sourceSet, Configuration
         String runtimeClasspathConfigurationName = sourceSet.getRuntimeClasspathConfigurationName();
         String sourceSetName = sourceSet.toString();
 
-        Configuration implementationConfiguration = configurations.maybeCreate(implementationConfigurationName);
+        Configuration implementationConfiguration = configurations.bucket(implementationConfigurationName, false);
         implementationConfiguration.setVisible(false);
         implementationConfiguration.setDescription("Implementation only dependencies for " + sourceSetName + ".");
-        implementationConfiguration.setCanBeConsumed(false);
-        implementationConfiguration.setCanBeResolved(false);
 
-        Configuration compileOnlyConfiguration = configurations.maybeCreate(compileOnlyConfigurationName);
+        Configuration compileOnlyConfiguration = configurations.bucket(compileOnlyConfigurationName, false);
         compileOnlyConfiguration.setVisible(false);
-        compileOnlyConfiguration.setCanBeConsumed(false);
-        compileOnlyConfiguration.setCanBeResolved(false);
         compileOnlyConfiguration.setDescription("Compile only dependencies for " + sourceSetName + ".");
 
-        ConfigurationInternal compileClasspathConfiguration = (ConfigurationInternal) configurations.maybeCreate(compileClasspathConfigurationName);
+        ConfigurationInternal compileClasspathConfiguration = configurations.resolvable(compileClasspathConfigurationName, false);
```

Comment:
Going from `maybeCreate` to a non-maybe create is a breaking change. I think we'll want to have a way to deprecate somebody creating the configuration before us (will be useful for `runtimeClasspath` below too), so that in 9.0 we can change this to a non-maybe create, and thus be sure we are creating these with the proper roles from the beginning.

Reply:
I don't think we have good options for setting up a deprecation warning to occur when someone creates a configuration we _might_ later create, as we won't know the name of the sourcesets they could create ahead of time.

We can check right ahead of our own `maybeCreate` calls here if the configuration already exists though, and emit the warning then.  This seems worthwhile, I'll implement it.

@big-guy - We talked about just using the intended roles here and removing  the `maybe()` calls, but after looking again and reconsidering I think Justin's right and we should go back to `maybe()`.

Reply:
Sticking with maybe here.